test/e2e/awscodecommitsource/awscodecommitsource.go

```diff
@@ -0,0 +1,206 @@
+		srcClient = f.DynamicClient.
+			Resource(conversion.V1alpha1GVR(apisources.AWSCodeCommitSourceResource)).
+			Namespace(f.Namespace.Name)
+
+		By("reading AWS credentials", func() {
+			sess := session.Must(session.NewSession())
+			ccClient = codecommit.New(sess)
+			awsCreds = readAWSCredentials(sess)
+		})
+
+		By("creating a Git repository", func() {
+			arn, undoCreateRepo := awscodecommit.CreateRepository(ccClient, f)
+			framework.AddCleanupAction(
+				undoCreateRepo,
+			)
+			repoARN = parseARN(arn)
+		})
+
+		By("creating an event sink", func() {
+			sink = sources.CreateEventDisplaySink(f.ClientSet, f.Namespace.Name)
+		})
+	})
+
+	It("should generate events on selected actions", func() {
+		By("creating an AWSCodeCommitSource object", func() {
+			src, err := createSource(srcClient, f.Namespace.Name, "test-", sink,
+				withARN(repoARN),
+				withCredentials(awsCreds),
+			)
+			Expect(err).ToNot(HaveOccurred())
+
+			sources.WaitForResourceReady(f.DynamicClient, src)
+		})
+
+		By("creating a Git commit", func() {
+			awscodecommit.CreateCommit(ccClient, repoARN.Resource)
+		})
+
+		// TODO(antoineco): check event-display logs, metrics?
+	})
+
+	It("should reject invalid specs", func() {
+		By("creating an AWSCodeCommitSource object with an invalid ARN", func() {
+			invalidARN := repoARN
+			invalidARN.AccountID = "invalid"
+
+			_, err := createSource(srcClient, f.Namespace.Name, "test-invalid-arn", sink,
+				withARN(invalidARN),
+				withCredentials(awsCreds),
+			)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("spec.arn: Invalid value: "))
+		})
+
+		By("creating an AWSCodeCommitSource object without credentials", func() {
+			_, err := createSource(srcClient, f.Namespace.Name, "test-nocreds", sink,
+				withARN(repoARN),
+			)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring(
+				`"spec.credentials.accessKeyID" must validate one and only one schema (oneOf).`))
+		})
+
+		By("creating an AWSCodeCommitSource object with invalid event types", func() {
+			_, err := createSource(srcClient, f.Namespace.Name, "test-invalid-eventtypes", sink,
+				withARN(repoARN),
+				withEventType("invalid"),
+				withCredentials(awsCreds),
+			)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring(`spec.eventTypes: Unsupported value: "invalid"`))
+		})
+	})
+})
+
+// createSource creates an AWSCodeCommitSource object initialized with the given options.
+func createSource(srcClient dynamic.ResourceInterface, namespace, namePrefix string,
+	sink duckv1.Destination, opts ...sourceOption) (*unstructured.Unstructured, error) {
+
+	src := &v1alpha1.AWSCodeCommitSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    namespace,
+			GenerateName: namePrefix,
+		},
+		Spec: v1alpha1.AWSCodeCommitSourceSpec{
+			Branch: awscodecommit.DefaultBranch,
+			SourceSpec: duckv1.SourceSpec{
+				Sink: sink,
+			},
+		},
+	}
+
+	for _, opt := range opts {
```

Comment:
what does this do ?

Reply:
You can pass functional options to the builder function to alter the contents of the object.

You can see an example of usage in the tests, where I call for example `withARN`,<`withEventTypes`, `withAWSCreds`, ...

It makes the function signature less crowded than requiring the caller to pass 6 or 7 parameters.